File: core/src/main/java/com/emarsys/core/util/log/entry/CrashLog.java
package com.emarsys.core.util.log.entry;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CrashLog implements LogEntry {

    private final Map<String, Object> data;

    public CrashLog(Throwable throwable) {
        data = new HashMap<>();
        if (throwable != null) {
            data.put("exception", throwable.getClass().getName());
            data.put("reason", throwable.getMessage());
            data.put("stack_trace", getStackTrace(throwable));
        }
    }

    @Override
    public String getTopic() {
        return "log_crash";
    }

    @Override
    public Map<String, Object> getData() {
        return data;
    }

    private List<String> getStackTrace(Throwable throwable) {
        StackTraceElement[] stackTrace = throwable.getStackTrace();
        int size = stackTrace.length;
        List<String> result = new ArrayList<>(size);

        for (StackTraceElement element : stackTrace) {
            result.add(element.toString());
        }

        return result;
    }

}

File: core/src/main/java/com/emarsys/core/util/log/entry/CrashLog.kt
package com.emarsys.core.util.log.entry

import java.util.*

class CrashLog(val throwable: Throwable?, additionalInformation: String? = null) : LogEntry {
    override val data: Map<String, Any?>
    override val topic: String
        get() = "log_crash"

    init {
        data = if (throwable != null) {
            mapOf(
                    "exception" to throwable.javaClass.name,
                    "reason" to throwable.message,
                    "additionalInformation" to additionalInformation,
                    "stackTrace" to getStackTrace(throwable)
            ).filterValues { it != null }
        } else {
            emptyMap()
        }
    }

    private fun getStackTrace(throwable: Throwable): List<String> {
        val stackTrace = throwable.stackTrace
        val size = stackTrace.size
        val result: MutableList<String> = ArrayList(size)
        for (element in stackTrace) {
            result.add(element.toString())
        }
        return result
    }
}